src/liballoc/collections/btree/node.rs

```diff
@@ -453,18 +481,29 @@ impl<'a, K, V, Type> NodeRef<marker::Mut<'a>, K, V, Type> {
 }
 
 impl<'a, K: 'a, V: 'a, Type> NodeRef<marker::Immut<'a>, K, V, Type> {
-    fn into_key_slice(self) -> &'a [K] {
-        unsafe { slice::from_raw_parts(MaybeUninit::first_ptr(&self.as_leaf().keys), self.len()) }
-    }
-
-    fn into_val_slice(self) -> &'a [V] {
-        unsafe { slice::from_raw_parts(MaybeUninit::first_ptr(&self.as_leaf().vals), self.len()) }
-    }
-
-    fn into_slices(self) -> (&'a [K], &'a [V]) {
+    /// SAFETY: the caller must ensure that the node has more than `idx` initialized elements.
+    unsafe fn into_key_at(self, idx: usize) -> &'a K {
+        debug_assert!(idx < self.len());
+        // SAFETY: we can access the entire node freely and do no need raw pointers,
+        // because there can be no mutable references to this immutable tree.
+        let leaf = unsafe { &(*self.as_leaf()) };
+        unsafe { leaf.keys.get_unchecked(idx).get_ref() }
+    }
+
+    /// SAFETY: the caller must ensure that the node has more than `idx` initialized elements.
+    unsafe fn into_val_at(self, idx: usize) -> &'a V {
+        debug_assert!(idx < self.len());
+        // SAFETY: we can access the entire node freely and do no need raw pointers,
+        // because there can be no mutable references to this immutable tree.
+        let leaf = unsafe { &(*self.as_leaf()) };
+        unsafe { leaf.vals.get_unchecked(idx).get_ref() }
+    }
+
+    /// SAFETY: the caller must ensure that the node has more than `idx` initialized elements.
+    unsafe fn into_key_val_at(self, idx: usize) -> (&'a K, &'a V) {
         // SAFETY: equivalent to reborrow() except not requiring Type: 'a
```

Comment:
I finally understand the meaning of this (Mark's) comment. Not yet why rustc starts requires the `'a` all over the place. But I wonder if it isn't more clear to make `Type: 'static` everywhere regardless? It doesn't look great but it makes it more clear that this is not a genuine type but an enumeration in the narrowest sense of the word.

Reply:
Why would `'static` be reasonable? Keys and values can contain references.

Reply:
I mean only for `Type` or in handle `NodeType`. Maybe also for `BorrowType`, but for some reason the borrow checker never ever complains about those.

Reply:
I have no idea what these are.^^ Better ask Mark about that.

Reply:
I don't know why we don't bind Type with 'static, doing so if needed indeed seems like not a problem to me.

Reply:
FYI, I think that the "requiring Type: 'a" is no more than a compiler artefact coming out before the real error about borrowing a from a temporary. There is no need for `reborrow` here, because we are a marker::Immut NodeRef, thus copyable. Without `reborrow`, I guess we are implicitly copied to a temporary, but then apparently the borrow checker has no problem returning a slice reference from that. Or it knows that both returned slice references stem from the same borrow or something. Anyway, over to another PR.